lib/internal/streams/pipeline.js

```diff
@@ -149,36 +152,38 @@ function pipeline(...streams) {
   let value;
   const destroys = [];
 
-  function finish(err, final) {
-    if (!error && err) {
+  let finishCount = 0;
+
+  function finish(err) {
+    const final = --finishCount === 0;
+
+    if (err && (!error || error.code === 'ERR_STREAM_PREMATURE_CLOSE')) {
       error = err;
     }
 
-    if (error || final) {
-      for (const destroy of destroys) {
-        destroy(error);
-      }
+    if (!error && !final) {
+      return;
+    }
+
+    for (const destroy of destroys) {
+      destroy(error);
     }
+    destroys.length = 0;
```

Comment:
I would prefer if we did a `destroys.shift()` instead.

Reply:
Fixed